src/github.com/getlantern/flashlight/app/app.go

```diff
@@ -0,0 +1,245 @@
+	exitCh      chan error
+	chExitFuncs chan func()
+}
+
+// Init initializes the App's state
+func (app *App) Init() {
+	app.exitCh = make(chan error, 1)
+	// use buffered channel to avoid blocking the caller of 'AddExitFunc'
+	// the number 10 is arbitrary
+	app.chExitFuncs = make(chan func(), 10)
+}
+
+// LogPanicAndExit logs a panic and then exits the application.
+func (app *App) LogPanicAndExit(msg string) {
+	cfg, err := config.Init(settings,
+		flashlight.PackageVersion,
+		app.Flags["configdir"].(string),
+		app.Flags["stickyconfig"].(bool),
+		app.Flags,
+	)
+	if err != nil {
+		panic("Error initializing config")
+	}
+	if err := logging.EnableFileLogging(); err != nil {
+		panic("Error initializing logging")
+	}
+
+	<-logging.Configure(eventual.DefaultGetter(""), "", cfg.Client.DeviceID, flashlight.Version, flashlight.RevisionDate)
+
+	log.Error(msg)
+
+	logging.Flush()
+	_ = logging.Close()
+
+	app.Exit(nil)
+}
+
+// Run starts the app. It will block until the app exits.
+func (app *App) Run() error {
+	// Run below in separate goroutine as config.Init() can potentially block when Lantern runs
+	// for the first time. User can still quit Lantern through systray menu when it happens.
+	go func() {
+		log.Debug(app.Flags)
+		if app.Flags["proxyall"].(bool) {
+			// If proxyall flag was supplied, force proxying of all
+			settings.SetProxyAll(true)
+		}
+
+		listenAddr := app.Flags["addr"].(string)
+		if listenAddr == "" {
+			listenAddr = "127.0.0.1:8787"
+		}
+
+		err := flashlight.Run(
+			listenAddr,
+			"127.0.0.1:8788",
+			app.Flags["configdir"].(string),
+			app.Flags["stickyconfig"].(bool),
+			settings.GetProxyAll,
+			app.Flags,
+			app.beforeStart,
+			app.afterStart,
+			app.onConfigUpdate,
+			settings,
+			app.Exit)
+		if err != nil {
+			app.Exit(err)
+			return
+		}
+	}()
+
+	return app.waitForExit()
+}
+
+func (app *App) beforeStart(cfg *config.Config) bool {
+	log.Debug("Got first config")
+
+	if cfg.CpuProfile != "" || cfg.MemProfile != "" {
+		log.Debugf("Start profiling with cpu file %s and mem file %s", cfg.CpuProfile, cfg.MemProfile)
+		finishProfiling := profiling.Start(cfg.CpuProfile, cfg.MemProfile)
+		app.AddExitFunc(finishProfiling)
+	}
+
+	if err := setUpPacTool(); err != nil {
+		app.Exit(err)
+	}
+
+	if app.Flags["clear-proxy-settings"].(bool) {
+		// This is a workaround that attempts to fix a Windows-only problem where
+		// Lantern was unable to clean the system's proxy settings before logging
+		// off.
+		//
+		// See: https://github.com/getlantern/lantern/issues/2776
+		log.Debug("Clearing proxy settings")
+		doPACOff(fmt.Sprintf("http://%s/proxy_on.pac", app.Flags["uiaddr"].(string)))
+		app.Exit(nil)
+	}
+
+	bootstrap, err := config.ReadBootstrapSettings()
+	var startupURL string
+	if err != nil {
+		log.Errorf("Could not read settings? %v", err)
+		startupURL = ""
+	} else {
+		startupURL = bootstrap.StartupUrl
+	}
+
+	log.Debugf("Starting client UI at %v", app.Flags["uiaddr"].(string))
```

Comment:
Another nit - seems like pulling `app.Flags["uiaddr"].(string)` out into a local variable is a little cleaner since we use it 3 times. Once in code plus a comment seems ok to me, but twice in code seems like too much. Not a big deal though.